### DISCUSSION TITLE
irtoy not showing up as option in "Hardware"

### DISCUSSION THREAD

**Discussion by pjotrek-b** [bengtmartensson/IrScrutinizer#519] [Hardware issues]
My irToy device does not show up as available hardware anymore.

I assume this is related to the "ir_toy" kernel driver (for lirc), because since then irToy registers as `/dev/lirc0` instead of the "good old" `/dev/ttyACM0`.

Here's a possibly related rollback-workaround: [RaspberryPI IrToy - restore /dev/ttyACM0 interface in newer kernels](https://gist.github.com/mnalis/2fed42a1abf8e98f45f6edf162bf609f).

Running irScrutinizer 2.4.1-SNAPSHOT (AppImage) on Xubuntu 20.04.6 (linux kernel 5.15.0-87)

Grateful for any input and help! :smile:

**Comment by bengtmartensson** [comment]
See [this thread](https://github.com/bengtmartensson/HarcHardware/discussions/25). (I am still very unhappy on how this was handled from the "Linux" side.)

The irtoy is currently supported using the /dev/lirc interface. (Sorry for Windows users...)
Can't you use that? (The /dev/lirc tab in hardware.) 

I will not support the linked kludge. _If_ you choose to install it, you can use irtoy with IrScrutinizer [version 2.3.1](https://github.com/bengtmartensson/IrScrutinizer/releases/tag/Version-2.3.1).

**Reply by pjotrek-b** [reply]
Thanks for the quick reply!

I've tried using the /dev/lirc0 interface in irScrutinizer, and I get the following error when trying to transmit a remote button (imported from a working lirc.conf remote file):

`HarcHardwareException: Hardware based exception: org.harctoolbox.devslashlirc.NotSupportedException setSendCarrier not supported by driver`

I totally get that you don't support the kludge (it's a wild hack).

**Reply by pjotrek-b** [reply]
lirc "irsend" works fine with the same device/setup.

**Comment by bengtmartensson** [comment]
Hmm, for me it is working (Fedora 39). What happens is that the [Linux irtoy driver](https://github.com/torvalds/linux/blob/master/drivers/media/rc/ir_toy.c) in your case says that it does not support setting carrier, but for me, it does support it. Anyhow, try [the new ci-build](https://github.com/torvalds/linux/blob/master/drivers/media/rc/ir_toy.c), which I hope will work for you.

See https://github.com/bengtmartensson/HarcHardware/issues/39

**Reply by pjotrek-b** [reply]
Thanks for checking!
Which kernel are you using?

**Comment by bengtmartensson** [comment]
> Which kernel are you using?

    Linux epsilon 6.7.5-200.fc39.x86_64 #1 SMP PREEMPT_DYNAMIC Sat Feb 17 17:20:08 UTC 2024 x86_64 GNU/Linux

The [relevant fix](https://github.com/torvalds/linux/commit/8bff1386d62d9406513a4806624d7cedefcb1ef9) was committed on 2021-09-30.

More importantly: the cu-build I linked to is supposed to fix the problem. also for "old" drivers. Did that work for you?

**Comment by pjotrek-b** [comment]
I am a bit confused how to quickly test a kernel module source code (you've linked), without going to build my own kernel from source?

Last time I've built my own kernels has been a while... :sweat_smile:

Am I overlooking or misunderstanding anything - or do you simply mean:
Build your own, and use one with `ir_toy.c` from later than 2021-09-30?

btw: I also wouldn't know what you mean by "ci/cu build"? (CI = continuous integration?)

Thanks for your time.
I'll update here, once I've managed to test a patched version.

**Comment by pjotrek-b** [comment]
May I ask, btw why you've removed (the existing) irToy support? :
Just curious.

I know irToy seems out-of-stock and "newer" Kernels have their own driver modules and (seem to) make fallback-access to the legacy /dev/ttyACM0 difficult/impossible, but...

I now can't run the same irScrutinizer (irS) version on both of my systems: one with kernel-module, the other one serial-device (legacy mode) :cry:

I haven't figured irS out quite so far, but seeing anything IR I shoot at it, transmitting signals quickly retrieved from almost any file-format - wow!

**Thank you! great work!** :clap: :cake:

**Comment by bengtmartensson** [comment]
> May I ask, btw why you've removed (the existing) irToy support? :
> Just curious.

I have answered [here](https://github.com/bengtmartensson/IrScrutinizer/discussions/411).

> btw: I also wouldn't know what you mean by "ci/cu build"?

The latter is a typo, should be "ci", which was previously linked.

I linked the critical commit mainly since I located it, and wanted to write it down.
